lms/services/http.py

```diff
@@ -0,0 +1,161 @@
+
+            This can be a (user, pass) 2-tuple which will be serialized using
+            HTTP Basic Authentication.
+
+            Or it can be a callable in order to implement other authentication
+            schemes. `requests` itself and libraries like `requests_oauthlib`
+            provide auth callables that you can use or you can write your own.
+
+            See: https://docs.python-requests.org/en/master/user/authentication/
+
+        :param timeout: How long (in seconds) to wait before raising an error.
+
+            This can be a (connect_timeout, read_timeout) 2-tuple or it can be
+            a single float that will be used as both the connect and read
+            timeout.
+
+            Note that the read_timeout is *not* a time limit on the entire
+            response download. It's a time limit on how long to wait *between
+            bytes from the server*. The entire download can take much longer.
+
+        :param schema: A schema class to use to validate the response
+        :type schema: lms.validation.RequestsResponseSchema
+
+        :param kwargs: Any other keyword arguments will be passed directly to
+            requests.Session().request():
+            https://docs.python-requests.org/en/latest/api/#requests.Session.request
+
+        :raise HTTPError: If sending the request or receiving the response
+            fails (DNS failure, refused connection, timeout, too many
+            redirects, etc).
+
+            The original exception from requests will be available as
+            HTTPError.__cause__.
+
+            In this case HTTPError.response will be None.
+
+        :raise HTTPError: If an error response (4xx or 5xx) is received.
+
+            The original exception from requests will be available as
+            HTTPError.__cause__.
+
+            The error response will be available as HTTPError.response.
+
+        :raise HTTPValidationError: If a `schema` argument is given and the
+            response fails validation with the schema.
+
+            The original lms.validation.ValidationError will be available as
+            HTTPValidationError.__cause__.
+
+            The invalid response will be available as
+            HTTPValidationError.response.
+        """
+        response = None
+
+        try:
+            response = self._session.request(
+                method,
+                url,
+                params=params,
+                data=data,
+                json=json,
+                headers=headers,
+                auth=auth,
+                timeout=timeout,
+                **kwargs,
+            )
+            response.raise_for_status()
+        except RequestException as err:
+            raise HTTPError(response) from err
+
+        if schema:
+            try:
+                response.validated_data = schema(response).parse()
+            except ValidationError as err:
+                raise HTTPValidationError(response) from err
```

Comment:
The schema argument isn't actually used by either `HAPI` or `LTIOutcomesClient`.

`HAPI` is sending requests to our own service so maybe we can more easily get away with just trusting the responses. But also I don't think it actually parses the response bodies anyway.

`LTIOutcomesClient` has had loads of customer-facing problems dealing with invalid or unexpected responses and would clearly benefit from better, schema-based validation. But that's beyond the scope of this PR. And LTI outcomes responses are XML so kind of a pain.